// fastify-type-provider-zod/dist/cjs/core.cjs
const require_errors = require("./errors.cjs");
const require_registry = require("./registry.cjs");
const require_utils = require("./utils.cjs");
const require_zod_to_json = require("./zod-to-json.cjs");
let zod_v4_core = require("zod/v4/core");
//#region src/core.ts
var isContentTypeResponse = (maybeSchema) => {
	return typeof maybeSchema === "object" && maybeSchema !== null && "content" in maybeSchema && typeof maybeSchema.content === "object" && maybeSchema.content !== null;
};
var defaultSkipList = [
	"/documentation/",
	"/documentation/initOAuth",
	"/documentation/json",
	"/documentation/uiConfig",
	"/documentation/yaml",
	"/documentation/*",
	"/documentation/static/*"
];
var createJsonSchemaTransform = ({ skipList = defaultSkipList, schemaRegistry = zod_v4_core.globalRegistry, zodToJsonConfig = {} }) => {
	const getIORegistries = require_registry.createIORegistriesProvider(schemaRegistry);
	return (document) => {
		require_utils.assertIsOpenAPIObject(document);
		const { schema, url } = document;
		if (!schema) return {
			schema,
			url
		};
		const { response, headers, querystring, body, params, hide, ...rest } = schema;
		const transformed = {};
		if (skipList.includes(url) || hide) {
			transformed.hide = true;
			return {
				schema: transformed,
				url
			};
		}
		const config = {
			target: require_utils.getJSONSchemaTarget(document.openapiObject.openapi),
			...zodToJsonConfig
		};
		const { inputRegistry, outputRegistry } = getIORegistries();
		const zodSchemas = {
			headers,
			querystring,
			body,
			params
		};
		for (const prop in zodSchemas) {
			const zodSchema = zodSchemas[prop];
			if (zodSchema) transformed[prop] = require_zod_to_json.zodSchemaToJson(zodSchema, inputRegistry, "input", config);
		}
		if (response) {
			transformed.response = {};
			for (const prop in response) {
				const responseSchema = response[prop];
				if (isContentTypeResponse(responseSchema)) {
					const responseObj = {};
					if (responseSchema.description) responseObj.description = responseSchema.description;
					responseObj.content = {};
					for (const [contentType, { schema: maybeSchema }] of Object.entries(responseSchema.content)) {
						const zodSchema = resolveSchema(maybeSchema);
						responseObj.content[contentType] = { schema: require_zod_to_json.zodSchemaToJson(zodSchema, outputRegistry, "output", config) };
					}
					transformed.response[prop] = responseObj;
					continue;
				}
				const zodSchema = resolveSchema(responseSchema);
				transformed.response[prop] = require_zod_to_json.zodSchemaToJson(zodSchema, outputRegistry, "output", config);
			}
		}
		for (const prop in rest) {
			const meta = rest[prop];
			if (meta) transformed[prop] = meta;
		}
		return {
			schema: transformed,
			url
		};
	};
};
var jsonSchemaTransform = createJsonSchemaTransform({});
var createJsonSchemaTransformObject = ({ schemaRegistry = zod_v4_core.globalRegistry, zodToJsonConfig = {} }) => (document) => {
	require_utils.assertIsOpenAPIObject(document);
	const config = {
		target: require_utils.getJSONSchemaTarget(document.openapiObject.openapi),
		...zodToJsonConfig
	};
	const { inputRegistry, outputRegistry } = require_registry.generateIORegistries(schemaRegistry);
	const inputSchemas = require_zod_to_json.zodRegistryToJson(inputRegistry, "input", config);
	const outputSchemas = require_zod_to_json.zodRegistryToJson(outputRegistry, "output", config);
	return {
		...document.openapiObject,
		components: {
			...document.openapiObject.components,
			schemas: {
				...document.openapiObject.components?.schemas,
				...inputSchemas,
				...outputSchemas
			}
		}
	};
};
var jsonSchemaTransformObject = createJsonSchemaTransformObject({});
var validatorCompiler = ({ schema }) => (data) => {
	const result = (0, zod_v4_core.safeParse)(schema, data);
	if (result.error) return { error: require_errors.createValidationError(result.error) };
	return { value: result.data };
};
function resolveSchema(maybeSchema) {
	if (maybeSchema instanceof zod_v4_core.$ZodType) return maybeSchema;
	if ("properties" in maybeSchema && maybeSchema.properties instanceof zod_v4_core.$ZodType) return maybeSchema.properties;
	throw new require_errors.InvalidSchemaError(JSON.stringify(maybeSchema));
}
var createSerializerCompiler = (options) => ({ schema: maybeSchema, method, url }) => {
	const schema = resolveSchema(maybeSchema);
	return (data) => {
		const result = (0, zod_v4_core.safeEncode)(schema, data);
		if (result.error) throw new require_errors.ResponseSerializationError(method, url, { cause: result.error });
		return JSON.stringify(result.data, options?.replacer);
	};
};
var serializerCompiler = createSerializerCompiler({});
//#endregion
exports.createJsonSchemaTransform = createJsonSchemaTransform;
exports.createJsonSchemaTransformObject = createJsonSchemaTransformObject;
exports.createSerializerCompiler = createSerializerCompiler;
exports.jsonSchemaTransform = jsonSchemaTransform;
exports.jsonSchemaTransformObject = jsonSchemaTransformObject;
exports.serializerCompiler = serializerCompiler;
exports.validatorCompiler = validatorCompiler;
