// fastify-type-provider-zod/dist/cjs/errors.cjs
const require_runtime = require("./_virtual/_rolldown/runtime.cjs");
let _fastify_error = require("@fastify/error");
_fastify_error = require_runtime.__toESM(_fastify_error, 1);
//#region src/errors.ts
var InvalidSchemaError = (0, _fastify_error.default)("FST_ERR_INVALID_SCHEMA", "Invalid schema passed: %s", 500);
var ZodFastifySchemaValidationErrorSymbol = Symbol.for("ZodFastifySchemaValidationError");
var ResponseSerializationBase = (0, _fastify_error.default)("FST_ERR_RESPONSE_SERIALIZATION", "Response doesn't match the schema", 500);
var ResponseSerializationError = class extends ResponseSerializationBase {
	method;
	url;
	constructor(method, url, options) {
		super({ cause: options.cause });
		this.method = method;
		this.url = url;
		this.cause = options.cause;
	}
};
function isResponseSerializationError(value) {
	return "method" in value;
}
function isZodFastifySchemaValidationError(error) {
	return typeof error === "object" && error !== null && error[ZodFastifySchemaValidationErrorSymbol] === true;
}
function hasZodFastifySchemaValidationErrors(error) {
	return typeof error === "object" && error !== null && "validation" in error && Array.isArray(error.validation) && error.validation.length > 0 && isZodFastifySchemaValidationError(error.validation[0]);
}
function omit(obj, keys) {
	const result = {};
	for (const key of Object.keys(obj)) if (!keys.includes(key)) result[key] = obj[key];
	return result;
}
function createValidationError(error) {
	return error.issues.map((issue) => {
		return {
			[ZodFastifySchemaValidationErrorSymbol]: true,
			keyword: issue.code,
			instancePath: `/${issue.path.join("/")}`,
			schemaPath: `#/${issue.path.join("/")}/${issue.code}`,
			message: issue.message,
			params: { ...omit(issue, [
				"path",
				"code",
				"message"
			]) }
		};
	});
}
//#endregion
exports.InvalidSchemaError = InvalidSchemaError;
exports.ResponseSerializationError = ResponseSerializationError;
exports.createValidationError = createValidationError;
exports.hasZodFastifySchemaValidationErrors = hasZodFastifySchemaValidationErrors;
exports.isResponseSerializationError = isResponseSerializationError;
