// fastify-type-provider-zod/dist/cjs/registry.cjs
let zod_v4_core = require("zod/v4/core");
//#region src/registry.ts
var getSchemaId = (id, io) => {
	return io === "input" ? `${id}Input` : id;
};
var WeakMapWithFallback = class extends WeakMap {
	fallback;
	constructor(fallback) {
		super();
		this.fallback = fallback;
	}
	get(key) {
		return super.get(key) ?? this.fallback.get(key);
	}
	has(key) {
		return super.has(key) || this.fallback.has(key);
	}
};
var copyRegistry = (baseRegistry, idReplaceFn) => {
	const copy = new zod_v4_core.$ZodRegistry();
	copy._map = new WeakMapWithFallback(baseRegistry._map);
	baseRegistry._idmap.forEach((schema, id) => {
		copy.add(schema, {
			...baseRegistry._map.get(schema),
			id: idReplaceFn(id)
		});
	});
	return copy;
};
var generateIORegistries = (baseRegistry) => {
	const inputRegistry = copyRegistry(baseRegistry, (id) => getSchemaId(id, "input"));
	const outputRegistry = copyRegistry(baseRegistry, (id) => getSchemaId(id, "output"));
	inputRegistry._idmap.forEach((_, id) => {
		if (outputRegistry._idmap.has(id)) throw new Error(`Collision detected for schema "${id}". An input and an output schema resolve to the same component name.`);
	});
	return {
		inputRegistry,
		outputRegistry
	};
};
/**
* Returns a memoized accessor for the derived input/output registries.
*
* The per-route transform runs once for every documented route, but the derived
* registries only depend on the contents of the base registry. Rebuilding them on
* every call is wasted work, so we cache the result and only recompute when the
* number of registered schemas changes (e.g. schemas added before a later
* `app.swagger()` call).
*/
var createIORegistriesProvider = (baseRegistry) => {
	let cache;
	return () => {
		const size = baseRegistry._idmap.size;
		if (!cache || cache.size !== size) cache = {
			size,
			registries: generateIORegistries(baseRegistry)
		};
		return cache.registries;
	};
};
//#endregion
exports.createIORegistriesProvider = createIORegistriesProvider;
exports.generateIORegistries = generateIORegistries;
