// fastify-type-provider-zod/dist/esm/core.js
import { InvalidSchemaError, ResponseSerializationError, createValidationError } from "./errors.js";
import { createIORegistriesProvider, generateIORegistries } from "./registry.js";
import { assertIsOpenAPIObject, getJSONSchemaTarget } from "./utils.js";
import { zodRegistryToJson, zodSchemaToJson } from "./zod-to-json.js";
import { $ZodType, globalRegistry, safeEncode, safeParse } from "zod/v4/core";
//#region src/core.ts
var isContentTypeResponse = (maybeSchema) => {
	return typeof maybeSchema === "object" && maybeSchema !== null && "content" in maybeSchema && typeof maybeSchema.content === "object" && maybeSchema.content !== null;
};
var defaultSkipList = [
	"/documentation/",
	"/documentation/initOAuth",
	"/documentation/json",
	"/documentation/uiConfig",
	"/documentation/yaml",
	"/documentation/*",
	"/documentation/static/*"
];
var createJsonSchemaTransform = ({ skipList = defaultSkipList, schemaRegistry = globalRegistry, zodToJsonConfig = {} }) => {
	const getIORegistries = createIORegistriesProvider(schemaRegistry);
	return (document) => {
		assertIsOpenAPIObject(document);
		const { schema, url } = document;
		if (!schema) return {
			schema,
			url
		};
		const { response, headers, querystring, body, params, hide, ...rest } = schema;
		const transformed = {};
		if (skipList.includes(url) || hide) {
			transformed.hide = true;
			return {
				schema: transformed,
				url
			};
		}
		const config = {
			target: getJSONSchemaTarget(document.openapiObject.openapi),
			...zodToJsonConfig
		};
		const { inputRegistry, outputRegistry } = getIORegistries();
		const zodSchemas = {
			headers,
			querystring,
			body,
			params
		};
		for (const prop in zodSchemas) {
			const zodSchema = zodSchemas[prop];
			if (zodSchema) transformed[prop] = zodSchemaToJson(zodSchema, inputRegistry, "input", config);
		}
		if (response) {
			transformed.response = {};
			for (const prop in response) {
				const responseSchema = response[prop];
				if (isContentTypeResponse(responseSchema)) {
					const responseObj = {};
					if (responseSchema.description) responseObj.description = responseSchema.description;
					responseObj.content = {};
					for (const [contentType, { schema: maybeSchema }] of Object.entries(responseSchema.content)) {
						const zodSchema = resolveSchema(maybeSchema);
						responseObj.content[contentType] = { schema: zodSchemaToJson(zodSchema, outputRegistry, "output", config) };
					}
					transformed.response[prop] = responseObj;
					continue;
				}
				const zodSchema = resolveSchema(responseSchema);
				transformed.response[prop] = zodSchemaToJson(zodSchema, outputRegistry, "output", config);
			}
		}
		for (const prop in rest) {
			const meta = rest[prop];
			if (meta) transformed[prop] = meta;
		}
		return {
			schema: transformed,
			url
		};
	};
};
var jsonSchemaTransform = createJsonSchemaTransform({});
var createJsonSchemaTransformObject = ({ schemaRegistry = globalRegistry, zodToJsonConfig = {} }) => (document) => {
	assertIsOpenAPIObject(document);
	const config = {
		target: getJSONSchemaTarget(document.openapiObject.openapi),
		...zodToJsonConfig
	};
	const { inputRegistry, outputRegistry } = generateIORegistries(schemaRegistry);
	const inputSchemas = zodRegistryToJson(inputRegistry, "input", config);
	const outputSchemas = zodRegistryToJson(outputRegistry, "output", config);
	return {
		...document.openapiObject,
		components: {
			...document.openapiObject.components,
			schemas: {
				...document.openapiObject.components?.schemas,
				...inputSchemas,
				...outputSchemas
			}
		}
	};
};
var jsonSchemaTransformObject = createJsonSchemaTransformObject({});
var validatorCompiler = ({ schema }) => (data) => {
	const result = safeParse(schema, data);
	if (result.error) return { error: createValidationError(result.error) };
	return { value: result.data };
};
function resolveSchema(maybeSchema) {
	if (maybeSchema instanceof $ZodType) return maybeSchema;
	if ("properties" in maybeSchema && maybeSchema.properties instanceof $ZodType) return maybeSchema.properties;
	throw new InvalidSchemaError(JSON.stringify(maybeSchema));
}
var createSerializerCompiler = (options) => ({ schema: maybeSchema, method, url }) => {
	const schema = resolveSchema(maybeSchema);
	return (data) => {
		const result = safeEncode(schema, data);
		if (result.error) throw new ResponseSerializationError(method, url, { cause: result.error });
		return JSON.stringify(result.data, options?.replacer);
	};
};
var serializerCompiler = createSerializerCompiler({});
//#endregion
export { createJsonSchemaTransform, createJsonSchemaTransformObject, createSerializerCompiler, jsonSchemaTransform, jsonSchemaTransformObject, serializerCompiler, validatorCompiler };
