// fastify-type-provider-zod/dist/esm/zod-to-json.js
import { getReferenceUri } from "./utils.js";
import { $ZodRegistry, $ZodType, toJSONSchema } from "zod/v4/core";
//#region src/zod-to-json.ts
var SCHEMA_REGISTRY_ID_PLACEHOLDER = "__SCHEMA__ID__PLACEHOLDER__";
var SCHEMA_URI_PLACEHOLDER = "__SCHEMA__PLACEHOLDER__";
/**
* Identity keywords that zod stamps on the root of a generated schema. They are not
* meaningful in the emitted document (we reference schemas through `$ref`), so they are
* removed for every target.
*/
var IDENTITY_KEYWORDS = [
	"id",
	"$id",
	"$schema"
];
/**
* Keywords that zod's `openapi-3.0` target can still emit (e.g. `contentEncoding` for
* `z.base64()`) but that are not part of the OpenAPI 3.0 schema subset. OpenAPI 3.1 /
* JSON Schema 2020-12 (`draft-2020-12`) does allow them, so they are only removed for 3.0.
*/
var OAS_3_0_INVALID_KEYWORDS = [
	"unevaluatedProperties",
	"dependentSchemas",
	"patternProperties",
	"propertyNames",
	"contentEncoding",
	"contentMediaType"
];
var getRemoveKeys = (target) => {
	return new Set(target === "openapi-3.0" ? [...IDENTITY_KEYWORDS, ...OAS_3_0_INVALID_KEYWORDS] : IDENTITY_KEYWORDS);
};
var SINGLE_SCHEMA_KEYWORDS = [
	"items",
	"additionalProperties",
	"not"
];
var SCHEMA_LIST_KEYWORDS = [
	"allOf",
	"anyOf",
	"oneOf"
];
/**
* Removes `removeKeys` from a generated JSON schema, recursing only into positions that
* actually hold subschemas (`properties` values, `items`, `additionalProperties`,
* `allOf`/`anyOf`/`oneOf`, `not`). It deliberately does not walk by key name everywhere:
* a user property called `id` lives as a key inside `properties` and must be preserved,
* while the `id` keyword on a schema object must be dropped.
*/
var sanitizeSchema = (value, removeKeys) => {
	if (Array.isArray(value)) return value.map((entry) => sanitizeSchema(entry, removeKeys));
	if (value === null || typeof value !== "object") return value;
	const result = {};
	for (const [key, child] of Object.entries(value)) {
		if (removeKeys.has(key)) continue;
		if (key === "properties" && child !== null && typeof child === "object") result[key] = Object.fromEntries(Object.entries(child).map(([name, schema]) => [name, sanitizeSchema(schema, removeKeys)]));
		else if (SINGLE_SCHEMA_KEYWORDS.includes(key)) result[key] = child !== null && typeof child === "object" ? sanitizeSchema(child, removeKeys) : child;
		else if (SCHEMA_LIST_KEYWORDS.includes(key)) result[key] = Array.isArray(child) ? child.map((entry) => sanitizeSchema(entry, removeKeys)) : child;
		else result[key] = child;
	}
	return result;
};
function isZodDate(entity) {
	return entity instanceof $ZodType && entity._zod.def.type === "date";
}
function isZodUnion(entity) {
	return entity instanceof $ZodType && entity._zod.def.type === "union";
}
function isZodUndefined(entity) {
	return entity instanceof $ZodType && entity._zod.def.type === "undefined";
}
/**
* The `openapi-3.0` target has no `null` type, so zod emits `z.null()` as
* `{ type: 'string', nullable: true, enum: [null] }`. Detect that branch so a nullable
* union can be collapsed back into a single `nullable: true` schema.
*/
var isNullBranch = (schema) => schema.nullable === true && Array.isArray(schema.enum) && schema.enum.length === 1 && schema.enum[0] === null;
var getOverride = (ctx, io, target) => {
	if (isZodUnion(ctx.zodSchema)) {
		ctx.jsonSchema.anyOf = ctx.jsonSchema.anyOf?.filter((schema) => Object.keys(schema).length > 0);
		const anyOf = ctx.jsonSchema.anyOf;
		if (target === "openapi-3.0" && anyOf) {
			const otherBranches = anyOf.filter((schema) => !isNullBranch(schema));
			if (otherBranches.length < anyOf.length && otherBranches.length > 0) if (otherBranches.length === 1) {
				delete ctx.jsonSchema.anyOf;
				Object.assign(ctx.jsonSchema, otherBranches[0], { nullable: true });
			} else {
				ctx.jsonSchema.anyOf = otherBranches;
				ctx.jsonSchema.nullable = true;
			}
		}
	}
	if (isZodDate(ctx.zodSchema)) {
		if (io === "output") {
			ctx.jsonSchema.type = "string";
			ctx.jsonSchema.format = "date-time";
		}
	}
	if (isZodUndefined(ctx.zodSchema)) {
		if (io === "output") ctx.jsonSchema.type = "null";
	}
};
var zodSchemaToJson = (zodSchema, registry, io, config) => {
	/**
	* Checks whether the provided schema is registered in the given registry.
	* If it is present and has an `id`, it can be referenced as component.
	*
	* @see https://github.com/turkerdev/fastify-type-provider-zod/issues/173
	*/
	const schemaRegistryEntry = registry.get(zodSchema);
	if (schemaRegistryEntry?.id) return { $ref: getReferenceUri(schemaRegistryEntry.id) };
	/**
	* Unfortunately, at the time of writing, there is no way to generate a schema with `$ref`
	* using `toJSONSchema` and a zod schema.
	*
	* As a workaround, we create a zod registry containing only the specific schema we want to convert.
	*
	* @see https://github.com/colinhacks/zod/issues/4281
	*/
	const tempRegistry = new $ZodRegistry();
	tempRegistry.add(zodSchema, { id: SCHEMA_REGISTRY_ID_PLACEHOLDER });
	const { schemas: { [SCHEMA_REGISTRY_ID_PLACEHOLDER]: result } } = toJSONSchema(tempRegistry, {
		...config,
		io,
		metadata: registry,
		unrepresentable: config.unrepresentable ?? "any",
		cycles: "ref",
		reused: "inline",
		/**
		* The uri option only allows customizing the base path of the `$ref`, and it automatically appends a path to it.
		* As a workaround, we set a placeholder that looks something like this.
		* @see https://github.com/colinhacks/zod/issues/4750
		*/
		uri: () => SCHEMA_URI_PLACEHOLDER,
		override: config.override ?? ((ctx) => getOverride(ctx, io, config.target))
	});
	/**
	* Remove identity/target-incompatible keywords first. This also drops the root `$id`
	* that zod sets to the uri placeholder, so the ref replacement below only ever sees the
	* placeholders that stand for real component references.
	*/
	const sanitized = sanitizeSchema(result, getRemoveKeys(config.target));
	/**
	* Replace the placeholder `$ref` values with the final component reference.
	*/
	return JSON.parse(JSON.stringify(sanitized), (_key, value) => typeof value === "string" && value.startsWith(SCHEMA_URI_PLACEHOLDER) ? getReferenceUri(value.slice(23)) : value);
};
var zodRegistryToJson = (registry, io, config) => {
	const result = toJSONSchema(registry, {
		...config,
		io,
		metadata: registry,
		unrepresentable: config.unrepresentable ?? "any",
		cycles: "ref",
		reused: "inline",
		uri: (id) => getReferenceUri(id),
		override: config.override ?? ((ctx) => getOverride(ctx, io, config.target))
	}).schemas;
	const removeKeys = getRemoveKeys(config.target);
	const jsonSchemas = {};
	for (const id in result) jsonSchemas[id] = sanitizeSchema(result[id], removeKeys);
	return jsonSchemas;
};
//#endregion
export { zodRegistryToJson, zodSchemaToJson };
